rpc.go

```diff
@@ -240,30 +239,30 @@ func (c *client) SendBatch(ctx context.Context, batch []hrpc.Call) (
 	if !ok {
 		return res, false
 	}
+	sendBatchSplitCount.Observe(float64(len(rpcByClient)))
 
 	// Send each group of RPCs to region client to be executed.
-	var (
-		wg sync.WaitGroup
-
-		mu   sync.Mutex
-		fail bool
-	)
-	wg.Add(len(rpcByClient))
+	type clientAndRPCs struct {
+		client hrpc.RegionClient
+		rpcs   []hrpc.Call
+	}
+	// keep track of the order requests are queued so that we can wait
+	// for their responses in the same order.
+	cAndRs := make([]clientAndRPCs, 0, len(rpcByClient))
 	for client, rpcs := range rpcByClient {
-		go func(client hrpc.RegionClient, rpcs []hrpc.Call) {
-			defer wg.Done()
-			client.QueueBatch(ctx, rpcs)
-			ctx, sp := observability.StartSpan(ctx, "waitForResult")
-			defer sp.End()
-			ok := c.waitForCompletion(ctx, client, rpcs, res, rpcToRes)
-			if !ok {
-				mu.Lock()
-				fail = true
-				mu.Unlock()
-			}
-		}(client, rpcs)
+		client.QueueBatch(ctx, rpcs)
+		cAndRs = append(cAndRs, clientAndRPCs{client, rpcs})
+	}
+
+	ctx, sp2 := observability.StartSpan(ctx, "waitForResult")
```

Comment:
Should this also be named something different (like `sp2`) because there's the initial `ctx` and `sp` in the deferred function at the beginning and could capture this `ctx` instead?

Reply:
Good catch. Jeez, I hate that this tracing stuff makes me have to think about the correct context to use...

I can move this code into it's own function to simplify things.

Reply:
Fixed.